// vidstack/dev/chunks/vidstack-CI_F5nTM.js
import { H as Host, g as effect, s as setAttribute } from './vidstack-mrj_B2Yo.js';
import { P as Poster } from './vidstack-D4cQThOl.js';
import './vidstack-CjVfv_MQ.js';
import './vidstack-CFPdVubu.js';
import './vidstack-pgEBqOw4.js';
import './vidstack-azSUCiDw.js';
import './vidstack-CnaYRoc3.js';
import './vidstack-xAXVpXZQ.js';
import './vidstack-Chu5mi6j.js';
import './vidstack-1gmLGa6x.js';
import './vidstack-DM_McBs5.js';
import './vidstack-ksPACRiU.js';

class MediaPosterElement extends Host(HTMLElement, Poster) {
  constructor() {
    super(...arguments);
    this._img = document.createElement("img");
  }
  static {
    this.tagName = "media-poster";
  }
  static {
    this.attrs = {
      crossOrigin: "crossorigin"
    };
  }
  onSetup() {
    this.$state.img.set(this._img);
  }
  onConnect() {
    const { src, alt, crossOrigin } = this.$state;
    if (this._img.parentNode !== this) {
      this.prepend(this._img);
    }
    effect(() => {
      setAttribute(this._img, "alt", alt());
      setAttribute(this._img, "crossorigin", crossOrigin());
      setAttribute(this._img, "src", src() || "");
    });
    effect(() => {
      const { loading, hidden } = this.$state;
      this._img.style.display = loading() || hidden() ? "none" : "";
    });
  }
}

export { MediaPosterElement };
