// vidstack/dev/chunks/vidstack-igu2AepS.js
import { C as Component, a as setStyle, q as onDispose, u as useContext } from './vidstack-mrj_B2Yo.js';
import { F as FocusVisibleController, t as tooltipContext } from './vidstack-D4cQThOl.js';
import { r as requestScopedAnimationFrame } from './vidstack-CjVfv_MQ.js';

class ControlsGroup extends Component {
  onAttach(el) {
    if (!el.style.pointerEvents)
      setStyle(el, "pointer-events", "auto");
  }
}

class TooltipTrigger extends Component {
  constructor() {
    super();
    new FocusVisibleController();
  }
  onConnect(el) {
    onDispose(
      requestScopedAnimationFrame(() => {
        if (!this.connectScope)
          return;
        this._attach();
        const tooltip = useContext(tooltipContext);
        onDispose(() => {
          const button = this._getButton();
          button && tooltip._detachTrigger(button);
        });
      })
    );
  }
  _attach() {
    const button = this._getButton(), tooltip = useContext(tooltipContext);
    button && tooltip._attachTrigger(button);
  }
  _getButton() {
    const candidate = this.el.firstElementChild;
    return candidate?.localName === "button" || candidate?.getAttribute("role") === "button" ? candidate : this.el;
  }
}

export { ControlsGroup as C, TooltipTrigger as T };
