// vidstack/prod/chunks/vidstack-BEFqUV5_.js
import { C as Component, a as setStyle, q as onDispose, u as useContext } from './vidstack-B11i_cNc.js';
import { F as FocusVisibleController, t as tooltipContext } from './vidstack-BPJwxG0c.js';
import { r as requestScopedAnimationFrame } from './vidstack-NiSULkLR.js';

class ControlsGroup extends Component {
  onAttach(el) {
    if (!el.style.pointerEvents)
      setStyle(el, "pointer-events", "auto");
  }
}

class TooltipTrigger extends Component {
  constructor() {
    super();
    new FocusVisibleController();
  }
  onConnect(el) {
    onDispose(
      requestScopedAnimationFrame(() => {
        if (!this.connectScope)
          return;
        this.xb();
        const tooltip = useContext(tooltipContext);
        onDispose(() => {
          const button = this.Bd();
          button && tooltip._e(button);
        });
      })
    );
  }
  xb() {
    const button = this.Bd(), tooltip = useContext(tooltipContext);
    button && tooltip.Ze(button);
  }
  Bd() {
    const candidate = this.el.firstElementChild;
    return candidate?.localName === "button" || candidate?.getAttribute("role") === "button" ? candidate : this.el;
  }
}

export { ControlsGroup as C, TooltipTrigger as T };
